src/v/coproc/script_context_frontend.cc

```diff
@@ -45,11 +45,34 @@ class high_offset_tracker {
 
 storage::log_reader_config get_reader(
   ss::abort_source& abort_src, read_context& rctx, const write_context& wctx) {
-    rctx.last_acked = wctx.offsets.empty() ? rctx.absolute_start
-                                           : wctx.min_offset();
+    auto offsets = wctx.unique_offsets();
+    model::offset start, end;
+    if (offsets.empty()) {
+        start = rctx.absolute_start;
+        end = rctx.input->last_stable_offset();
+    } else if (offsets.size() == 1) {
+        start = *offsets.begin();
+        end = rctx.input->last_stable_offset();
+    } else {
+        /// The fact that this case was entered means a retry is to be performed
+        /// since there is a difference in where all output topics stand.
+        start = *offsets.begin();
+        /// Limit the read to process only up until the next output topic. This
+        /// way \ref last_read isn't ahead of any logs which would perform a
+        /// promotion after the response is processed in
+        /// script_context_backend.cc
+        end = *(++offsets.begin()) - model::offset{1};
```

Comment:
Going to pre-emptively explain this odd line for other reviewers. The following line:
```
    ctx->rctx.last_read = info.last + model::offset{1};
```
Is called after the read is performed. Its intention is to store the offset of the current read. It really should be named `next_read` because it adds one the the observed offset.

So when calculating the endpoint here on line 64, the -1 is to account for this. This line really wants to mean "stop reading at `last_read` not `next_read`"